File: src/csvParser.js
/**
 * CSV Parser Module
 * Parses CSV files containing flashcard data (question, answer pairs)
 */

/**
 * Parses CSV content into an array of flashcard objects
 * @param {string} csvContent - The raw CSV content as a string
 * @returns {Array<{id: number, question: string, answer: string}>} Array of flashcard objects
 * @throws {Error} If CSV structure is invalid
 */
export function parseCSV(csvContent) {
  // Validate input
  if (typeof csvContent !== 'string') {
    throw new Error('CSV content must be a string');
  }

  // Handle empty content
  if (csvContent.trim() === '') {
    throw new Error('The selected file is empty. Please choose a file with flashcard data.');
  }

  // Split into rows, handling different line endings
  const rows = csvContent.split(/\r?\n/);
  
  const flashcards = [];
  let id = 1;

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i].trim();
    
    // Skip empty rows
    if (row === '') {
      continue;
    }

    // Parse the row - handle basic CSV parsing
    const columns = parseCSVRow(row);

    // Validate structure - must have at least 2 columns
    if (columns.length < 2) {
      throw new Error('Invalid CSV format. Please ensure the file has two columns: question and answer.');
    }

    const question = columns[0].trim();
    const answer = columns[1].trim();

    // Skip rows where both question and answer are empty
    if (question === '' && answer === '') {
      continue;
    }

    // Create flashcard object
    flashcards.push({
      id: id++,
      question: question,
      answer: answer
    });
  }

  // Check if we got any valid flashcards
  if (flashcards.length === 0) {
    throw new Error('The selected file is empty. Please choose a file with flashcard data.');
  }

  return flashcards;
}

/**
 * Parses a single CSV row, handling quotes and special characters
 * @param {string} row - A single row from the CSV
 * @returns {Array<string>} Array of column values
 */
function parseCSVRow(row) {
  const columns = [];
  let currentColumn = '';
  let insideQuotes = false;
  
  for (let i = 0; i < row.length; i++) {
    const char = row[i];
    const nextChar = row[i + 1];

    if (char === '"') {
      if (insideQuotes && nextChar === '"') {
        // Escaped quote - add a single quote to the column
        currentColumn += '"';
        i++; // Skip the next quote
      } else {
        // Toggle quote state
        insideQuotes = !insideQuotes;
      }
    } else if (char === ',' && !insideQuotes) {
      // Column separator - save current column and start new one
      columns.push(currentColumn);
      currentColumn = '';
    } else {
      // Regular character - add to current column
      currentColumn += char;
    }
  }

  // Add the last column
  columns.push(currentColumn);

  return columns;
}

/**
 * Validates CSV structure without parsing the entire content
 * @param {string} csvContent - The raw CSV content as a string
 * @returns {boolean} True if structure is valid, false otherwise
 */
export function validateCSVStructure(csvContent) {
  try {
    parseCSV(csvContent);
    return true;
  } catch (error) {
    return false;
  }
}

/**
 * Serializes flashcards back to CSV format (for testing round-trip)
 * @param {Array<{question: string, answer: string}>} flashcards - Array of flashcard objects
 * @returns {string} CSV formatted string
 */
export function serializeToCSV(flashcards) {
  return flashcards.map(card => {
    // Escape quotes and wrap in quotes if necessary
    const question = escapeCSVField(card.question);
    const answer = escapeCSVField(card.answer);
    return `${question},${answer}`;
  }).join('\n');
}

/**
 * Escapes a CSV field by wrapping in quotes if it contains special characters
 * @param {string} field - The field to escape
 * @returns {string} Escaped field
 */
function escapeCSVField(field) {
  // If field contains comma, quote, or newline, wrap in quotes and escape internal quotes
  if (field.includes(',') || field.includes('"') || field.includes('\n')) {
    return `"${field.replace(/"/g, '""')}"`;
  }
  return field;
}

File: utils/csv_parser.py
"""CSV parser module for loading flashcard data from CSV files."""

import csv
from typing import List
from io import StringIO
from models.flashcard import Flashcard


class CSVParseError(Exception):
    """Exception raised when CSV parsing fails."""
    pass


def parse_csv(csv_content: str) -> List[Flashcard]:
    """Parse CSV content into a list of Flashcard objects.
    
    This function parses CSV content where the first column is the question
    and the second column is the answer. It handles UTF-8 encoding, filters
    out empty rows, and validates the CSV structure.
    
    Args:
        csv_content: String containing CSV data with questions and answers
        
    Returns:
        List of Flashcard objects parsed from the CSV content
        
    Raises:
        CSVParseError: If the CSV is malformed or doesn't have at least 2 columns
        
    Examples:
        >>> csv_data = "What is 2+2?,4\\nWhat is Python?,A programming language"
        >>> cards = parse_csv(csv_data)
        >>> len(cards)
        2
    """
    if not csv_content or not csv_content.strip():
        return []
    
    flashcards = []
    csv_file = StringIO(csv_content)
    
    try:
        # Use csv.reader to handle edge cases like quotes, commas in fields, etc.
        reader = csv.reader(csv_file)
        
        for row_num, row in enumerate(reader, start=1):
            # Skip empty rows (rows with no content or only whitespace)
            if not row or all(not cell.strip() for cell in row):
                continue
            
            # Validate that the row has at least 2 columns
            if len(row) < 2:
                raise CSVParseError(
                    f"Row {row_num} has fewer than 2 columns. "
                    f"CSV must have at least two columns for questions and answers."
                )
            
            # Extract question and answer (first two columns)
            question = row[0].strip()
            answer = row[1].strip()
            
            # Skip rows where both question and answer are empty
            if not question and not answer:
                continue
            
            # Create flashcard with 1-based ID
            flashcard = Flashcard(
                id=len(flashcards) + 1,
                question=question,
                answer=answer
            )
            flashcards.append(flashcard)
    
    except csv.Error as e:
        raise CSVParseError(f"Failed to parse CSV: {str(e)}")
    
    return flashcards


def validate_csv_structure(csv_content: str) -> bool:
    """Validate that CSV content has the proper structure.
    
    Checks that the CSV has at least 2 columns in each non-empty row.
    
    Args:
        csv_content: String containing CSV data to validate
        
    Returns:
        True if the CSV structure is valid, False otherwise
    """
    if not csv_content or not csv_content.strip():
        return True  # Empty content is valid (will result in empty deck)
    
    try:
        csv_file = StringIO(csv_content)
        reader = csv.reader(csv_file)
        
        for row in reader:
            # Skip empty rows
            if not row or all(not cell.strip() for cell in row):
                continue
            
            # Check if row has at least 2 columns
            if len(row) < 2:
                return False
        
        return True
    
    except csv.Error:
        return False


def load_csv_file(filepath: str) -> List[Flashcard]:
    """Load and parse a CSV file from the filesystem.
    
    Reads the file with UTF-8 encoding to support international characters
    (e.g., Chinese text) and parses it into Flashcard objects.
    
    Args:
        filepath: Path to the CSV file to load
        
    Returns:
        List of Flashcard objects parsed from the file
        
    Raises:
        FileNotFoundError: If the file doesn't exist
        PermissionError: If the file cannot be read
        CSVParseError: If the CSV content is malformed
        UnicodeDecodeError: If the file is not UTF-8 encoded
    """
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            csv_content = f.read()
        
        return parse_csv(csv_content)
    
    except FileNotFoundError:
        raise FileNotFoundError(f"CSV file not found: {filepath}")
    
    except PermissionError:
        raise PermissionError(f"Permission denied reading file: {filepath}")
    
    except UnicodeDecodeError as e:
        raise UnicodeDecodeError(
            e.encoding, e.object, e.start, e.end,
            f"Unable to read file. Please ensure the file is UTF-8 encoded."
        )
